app/vmselect/promql/exec.go

```diff
@@ -33,16 +104,40 @@ func Exec(ec *EvalConfig, q string, isFirstPointOnly bool) ([]netstorage.Result,
 		}()
 	}
 
+	stopCh := make(chan error, 1)
+	// TODO(dexter): check performance
+	resultCh := make(chan []netstorage.Result)
+	c := runningQueries.Add(Query{
+		q:       q,
+		ec:      ec,
+		startAt: time.Now(),
+		stopCh:  stopCh,
+	})
+	defer runningQueries.Delete(c)
+
+	go exec(ec, q, isFirstPointOnly, stopCh, resultCh)
+	for {
+		select {
+		case err := <-stopCh:
+			logger.Infof(err.Error())
+			return nil, err
+		case result := <-resultCh:
+			return result, nil
+		}
+	}
+}
+
+func exec(ec *EvalConfig, q string, isFirstPointOnly bool, stopCh chan error, resultCh chan []netstorage.Result) {
 	ec.validate()
 
 	e, err := parsePromQLWithCache(q)
 	if err != nil {
-		return nil, err
+		stopCh <- err
```

Comment:
It looks like `return` is missing after such lines

Reply:
I use channel for returning result or errors, that's why `return` is missing.